cert/aziot-certd/src/lib.rs

```diff
@@ -195,654 +242,356 @@ impl UpdateConfig for Api {
+                    .append_entry_by_text(name, value)
+                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
+            }
     }
+
+    Ok(builder.build())
 }
 
-fn create_cert<'a>(
+#[async_recursion]
+async fn create_cert_inner<'a>(
     api: &'a mut Api,
     id: &'a str,
-    csr: &'a [u8],
-    issuer: Option<(&'a str, &'a aziot_key_common::KeyHandle)>,
-) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<Vec<u8>, Error>> + Send + 'a>> {
-    // Creating a cert is recursive in some cases. An async fn cannot recurse because its RPIT Future type would end up being infinitely sized,
-    // so it needs to be boxed. So we have a non-async fn returning a boxed future, where the future being boxed is the result of an inner asyn fn,
-    // and the recursive call is for the outer boxed-future-returning fn.
-
-    async fn create_cert_inner(
-        api: &mut Api,
-        id: &str,
-        csr: &[u8],
-        issuer: Option<(&str, &aziot_key_common::KeyHandle)>,
-    ) -> Result<Vec<u8>, Error> {
-        // Look up issuance options for this certificate ID.
-        let cert_options = api.cert_issuance.certs.get(id);
-
-        if let Some((issuer_id, issuer_private_key)) = issuer {
-            // Issuer is explicitly specified, so load it and use it to sign the CSR.
-
-            let x509_req = openssl::x509::X509Req::from_pem(csr)
-                .map_err(|err| Error::invalid_parameter("csr", err))?;
-            let x509_req_public_key = x509_req
-                .public_key()
-                .map_err(|err| Error::invalid_parameter("csr", err))?;
-            if !x509_req
-                .verify(&x509_req_public_key)
-                .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?
-            {
-                return Err(Error::invalid_parameter(
-                    "csr",
-                    "CSR failed to be verified with its public key",
-                ));
-            }
-
-            // If issuance options are not provided for this certificate ID, use defaults.
-            let mut expiry_days = 30;
-            let mut subject_name = x509_req.subject_name();
-            let common_name;
-
-            if let Some(options) = cert_options {
-                if let Some(d) = options.expiry_days {
-                    expiry_days = d;
-                }
-
-                if let Some(c) = &options.common_name {
-                    let mut name_builder = openssl::x509::X509Name::builder()
-                        .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-
-                    name_builder
-                        .append_entry_by_text("CN", c)
-                        .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-                    common_name = name_builder.build();
-                    subject_name = &common_name;
+    csr: (&'a X509Req, &'a PKey<Public>),
```

Comment:
Since the first thing you do with it is destructure it:

```suggestion
    (req, pubkey): (&'a X509Req, &'a PKeyRef<Public>),
```

Reply:
Fixed in e2d57edbe5898347cf7633f3a59c841737c9e963.